src/nn/transformer/base.rs

```diff
@@ -0,0 +1,221 @@
+    for MultiHeadAttention<M, N, K, V, H>
+{
+    fn reset_params<R: Rng>(&mut self, rng: &mut R) {
+        self.w_q.reset_params(rng);
+        self.w_k.reset_params(rng);
+        self.w_v.reset_params(rng);
+        self.w_o.reset_params(rng);
+    }
+}
+
+impl<const M: usize, const N: usize, const K: usize, const V: usize, const H: usize>
+    CanUpdateWithGradients for MultiHeadAttention<M, N, K, V, H>
+{
+    fn update<G: GradientProvider>(&mut self, grads: &mut G) {
+        self.w_q.update(grads);
+        self.w_k.update(grads);
+        self.w_v.update(grads);
+        self.w_o.update(grads);
+    }
+}
+
+/// Normal self attention (where same tensors are used for keys, queries and values)
+impl<const M: usize, const K: usize, const V: usize, const S: usize, const H: usize, T: Tape>
+    Module<Tensor2D<S, M, T>> for MultiHeadAttention<M, M, K, V, H>
+where
+    Assert<{ S * K == H * S * (K / H) }>: ConstTrue,
+    Assert<{ S * V == H * S * (V / H) }>: ConstTrue,
+    Assert<{ H * S * (V / H) == S * V }>: ConstTrue,
+{
+    type Output = Tensor2D<S, M, T>;
+
+    fn forward(&self, input: Tensor2D<S, M, T>) -> Self::Output {
+        let (input, tape) = input.split_tape();
+        let (queries, tape) = Reshape::<Tensor3D<H, S, { K / H }, T>>::reshape(
+            self.w_q.forward(input.duplicate().put_tape(tape)),
+        )
+        .split_tape();
+        let (keys, tape) = Reshape::<Tensor3D<H, S, { K / H }, T>>::reshape(
+            self.w_k.forward(input.duplicate().put_tape(tape)),
+        )
+        .split_tape();
+        let (values, tape) = Reshape::<Tensor3D<H, S, { V / H }, T>>::reshape(
+            self.w_v.forward(input.put_tape(tape)),
+        )
+        .split_tape();
+
+        // Get weights
+        let token_weights = batch_3d_matmul_transpose(queries.put_tape(tape), &keys) / (M as f32);
+
+        // Softmax on last dimension
+        let token_weights = softmax(token_weights);
+
+        // Get new tokens
+        let tokens: Tensor2D<S, V, T> = batch_3d_matmul(token_weights, &values).reshape();
+        self.w_o.forward(tokens)
+    }
+}
+
+/// Encoder-Decoder style self attention where one set of tensors is used for values and keys, and another is used for queries
+impl<
+        const M: usize,
+        const N: usize,
+        const K: usize,
+        const V: usize,
+        const S1: usize,
+        const S2: usize,
+        const H: usize,
+        T: Tape,
+    > Module<(Tensor2D<S1, M, T>, Tensor2D<S2, N, T>)> for MultiHeadAttention<M, N, K, V, H>
+where
+    Assert<{ S1 * K == H * S1 * (K / H) }>: ConstTrue,
+    Assert<{ S2 * K == H * S2 * (K / H) }>: ConstTrue,
+    Assert<{ S2 * V == H * S2 * (V / H) }>: ConstTrue,
+    Assert<{ H * S1 * (V / H) == S1 * V }>: ConstTrue,
+{
+    type Output = (Tensor2D<S1, M, T>, T);
+
+    fn forward(&self, (input, from_enc): (Tensor2D<S1, M, T>, Tensor2D<S2, N, T>)) -> Self::Output {
```

Comment:
So generally for tape, there should only be 1 on the input & one on the output. And at least 1 tensor in every operation in forward needs to own `T`. so all the reshape ops should have tape, and then the tape needs to be in  `queries` when input to the matmul argument.

Then the output of this can just be `type Output = Tensor2D<S1, M, T>`, as that output tensor owns `T`, so its not separate